### ISSUE TITLE
plans for sourmash 3.0 and sourmash 4.0

### ISSUE THREAD

**Issue by ctb** [dib-lab/sourmash#762]
Results of a brief chat with @luizirber, my version:

sourmash 3.0 will be the first release of an oxidized sourmash (i.e. conversion from C++ to Rust). This can happen soon (I think we're mostly waiting for me to finish #556).

Notes:
* I'm worried about installation / execution problems, but bioconda doesn't respect alpha labels, so we can't release a 3.0alpha and ask people to try it out. So, instead, we're going to release a sourmash 3.0alpha and modify our default install instructions to suggest installing sourmash==2.x.
* sourmash 3.0 would remain mostly (entirely?) command-line compatible with sourmash 2.x.
* @luizirber wants to start using a deprecation decorator or warning to flag upcoming removals and name changes.

---

sourmash 4.0 would then a reasonably incompatible update of sourmash to do things like change/update signature formats, introduced `--scaled` as default parameter instead of regular MinHash, etc.

Notes:
* I would like to do a better job of simulations and theory before we release sourmash 4.0, or at least get it on the radar. We need to start understanding (and explicating) where the basic scaled approach is good and not so good.

----

@luizirber does this capture your thoughts?

@olgabot do you and yours have any thoughts? Note, in terms of releasing incompatibilities, I'd prefer to bump major version number (3.x, 4.x, 5.x) faster, than be slower/careful about major releases.

**Comment by luizirber** [comment]
> Notes:
> 
>     * I'm worried about installation / execution problems, but bioconda doesn't respect alpha labels, so we can't release a 3.0alpha and ask people to try it out. So, instead, we're going to release a sourmash 3.0alpha and modify our default install instructions to suggest installing sourmash==2.x.

another way is to create a temporary package (call it `sourmash3`?) to try out the new build system and request comments before releasing it in the `sourmash` recipe. I'll open an issue in bioconda and see what they say.

For pypi, we can also create `sourmash3`, test it out, and then [deprecate](https://stackoverflow.com/a/55165236) the package and post instructions to use regular `sourmash`?

>     * sourmash 3.0 would remain mostly (entirely?) command-line compatible with sourmash 2.x.

"Entirely" is the goal, also at the Python API level.

> sourmash 4.0 would then a reasonably incompatible update of sourmash to do things like change/update signature formats, introduced `--scaled` as default parameter instead of regular MinHash, etc.

I think we can change/update signature formats without breaking anything, just like SBTs continue working after being upgraded. But definitely the version `1` for minhash can go in `4.0`, it will take some tweaking to get right and `3.0` is already pretty crowded...

> @luizirber does this capture your thoughts?

Yup, thanks for writing it down!

**Comment by olgabot** [comment]
> sourmash 3.0 will be the first release of an oxidized sourmash (i.e. conversion from C++ to Rust). This can happen soon (I think we're mostly waiting for me to finish #556).
Whoo exciting!!


> * I'm worried about installation / execution problems, but bioconda doesn't respect alpha labels, so we can't release a 3.0alpha and ask people to try it out. So, instead, we're going to release a sourmash 3.0alpha and modify our default install instructions to suggest installing sourmash==2.x.
Does the alpha version **need** to be on bioconda? Would it be okay to just be on PyPI in the meantime?

> * sourmash 3.0 would remain mostly (entirely?) command-line compatible with sourmash 2.x.
Sounds good to me!

> * @luizirber wants to start using a deprecation decorator or warning to flag upcoming removals and name changes.
Yes, deprecation warnings would help.

> sourmash 4.0 would then a reasonably incompatible update of sourmash to do things like change/update signature formats, introduced `--scaled` as default parameter instead of regular MinHash, etc.

Not sure I understand what this means: "`--scaled` as the default parameter", could you elaborate? I usually use a "flat rate" aka just the `--num-hashes` parameter.

> @olgabot do you and yours have any thoughts? Note, in terms of releasing incompatibilities, I'd prefer to bump major version number (3.x, 4.x, 5.x) faster, than be slower/careful about major releases.


@pranathivemuri @phoenixaja @snafees -- do you have any input?

**Comment by pranathivemuri** [comment]
not input, but a question. so, we would have both python and rust versions of sourmash cli api to support and use after version 3.0 is released? how would rust be better or worse than python? can we document that?

**Comment by luizirber** [comment]
> not input, but a question. so, we would have both python and rust versions of sourmash cli api to support and use after version 3.0 is released? 

Python version is the only supported CLI version for 3.0 (just as it is now). There is a Rust CLI lurking in the repo, but it is not official nor supported. I mostly use it for testing out compatibility for the things that are not exposed to Python yet (like search and SBTs) and experiment with ideas.

> how would rust be better or worse than python?

They are complementary, just as Python and C++ are complementary now. Rust allows better low-overhead parallelism in the future, but this is not exposed to Python yet.

One thing that's way faster is signature loading, especially with very large metagenomics signatures. This is in #532 and will be merged for 3.0.

> can we document that?

Yup... are the dev docs a good place? We have a bunch of conversations scattered in issues here, but there isn't a clear entrypoint to the discussion.

**Comment by pranathivemuri** [comment]
I think dev docs is a good place. Thanks for answering my questions!

**Comment by ctb** [comment]
On Sun, Nov 03, 2019 at 06:44:00PM -0800, Olga Botvinnik wrote:
> > sourmash 4.0 would then a reasonably incompatible update of sourmash to do things like change/update signature formats, introduced `--scaled` as default parameter instead of regular MinHash, etc.
> 
> Not sure I understand what this means: "`--scaled` as the default parameter", could you elaborate? I usually use a "flat rate" aka just the `--num-hashes` parameter.

Right, our use for all the metagenomics-y things is almost always --scaled.

**Comment by luizirber** [comment]
See https://github.com/bioconda/bioconda-recipes/issues/18659 for a way to do a sourmash 3.0 prerelease without bumping the bioconda recipe

**Comment by luizirber** [comment]
Proposal: start a `2.x` branch from the current `master`, and merge #424.

All current PRs can still point to 2.x, but new PRs will be merged into master. The goal is to have **`2.x` as the long-running parallel branch, instead of [`core/rust`](https://github.com/dib-lab/sourmash/compare/core/rust)**.

The drawback is that we well have a small time window (before 3.x is released) of features on both branches. To be clear, this already happens because I need to port all changes in `master` to `core/rust`, but no one sees it. I'll continue to port current PRs merged into `2.x`  into `master`, but all new PRs will default to `master`.

**Comment by luizirber** [comment]
I was going thru open PRs, and I was thinking about merging these before #424 
- [ ] [hash values to kmers script](https://github.com/dib-lab/sourmash/pull/724)
- [ ] [hp encoding](https://github.com/dib-lab/sourmash/pull/758)

And then this after #424, but before 3.x:
- [ ] [the Index abc](https://github.com/dib-lab/sourmash/pull/556)

After the #424 merge also switch around branches, based on previous message.

Thoughs?

**Comment by pranathivemuri** [comment]
I am fine with merging the hp encoding in python, but a thing to note is it's not implemented in rust yet

**Comment by luizirber** [comment]
> I am fine with merging the hp encoding in python, but a thing to note is it's not implemented in rust yet

That's OK, I think it's better to merge it and then fix #424 than the other way around

**Comment by luizirber** [comment]
for 4.0: follow @standage hints in https://github.com/dib-lab/sourmash/pull/785 and deprecate some subcommands

**Comment by bluegenes** [comment]
>Not sure I understand what this means: "--scaled as the default parameter", could you elaborate? I usually use a "flat rate" aka just the --num-hashes parameter.

@olgabot follow up to your `--scaled` question, mostly so the details are here for others w/same q:

Since scaled hashes retain the same fraction (rather than number) of kmer hashes, they compress large and small datasets at the same rate. This results in a more representative set of hashes for comparing datasets of disparate sizes, but can sacrifice some of the memory and storage benefits of standard MinHash techniques, as the signature size scales with the number of unique kmers rather than remaining fixed. 

Use of the "scaled" factor enables user-modification of the trade-off between search precision and sketch size, with the caveat that searches and comparisons can only be conducted using signatures generated with identical "scaled" values (compressed at the same rate).

**Comment by luizirber** [comment]
#424 landed, and is in the current master branch.

I'm working on #532 next, and then packaging in bioconda (with an `-rc` candidate version?)

**Comment by ctb** [comment]
@olgabot @bluegenes re `--scaled`, I actually documented that! https://sourmash.readthedocs.io/en/latest/api-example.html#sourmash-minhash-objects-and-manipulations

**Comment by luizirber** [comment]
Proposal: after merging #797 and #808 do a `3.0.0-rc1` release, and upload it to PyPI. This will get picked up by bioconda, and we can work on fixing the build process (I tried it in https://github.com/bioconda/bioconda-recipes/pull/19406 but I don't think the build system likes changing URLs like that).

Leave `3.0.0-rc1` up for about a week, and then release as `3.0.0` if nothing catastrophic shows up.

**Comment by ctb** [comment]
On Sun, Dec 22, 2019 at 03:16:02PM -0800, Luiz Irber wrote:
> Proposal: after merging #797 and #808 do a `3.0.0-rc1` release, and upload it to PyPI. This will get picked up by bioconda, and we can work on fixing the build process (I tried it in https://github.com/bioconda/bioconda-recipes/pull/19406 but I don't think the build system likes changing URLs like that).
> 
> Leave `3.0.0-rc1` up for about a week, and then release as `3.0.0` if nothing catastrophic shows up.

I thought bioconda didn't allow rc releases? Were we going to do the
"create our own channel and check it that way" thing?

**Comment by luizirber** [comment]
> I thought bioconda didn't allow rc releases? Were we going to do the "create our own channel and check it that way" thing?

They do accept prereleases (we were `2.0.0a1367`, after all =]). The issue with conda is that as soon as there is an `3.0.0rc1` package, it is going to be installed by default when people do `conda install sourmash`.

**Comment by luizirber** [comment]
`3.0.0rc1` in PyPI, twitter announcement:
https://twitter.com/ctitusbrown/status/1210635545077927937

**Comment by luizirber** [comment]
Bioconda build seems to be working, but at this point thinking about holding the merge for tomorrow and doing `3.0.0` directly (instead of adding `3.0.0rc1`?)
https://github.com/bioconda/bioconda-recipes/pull/19502

**Comment by ctb** [comment]
+1

**Comment by luizirber** [comment]
sourmash 3.0.0 released: https://pypi.org/project/sourmash/3.0.0/
waiting for macOS build on bioconda to finish now, and then annouce it.

**Comment by luizirber** [comment]
For 3.1: merge #785 and #532 (probably in this order)

**Comment by ctb** [comment]
Closing this issue now that 3.0 is out, although @luizirber is welcome to re-open it if he thinks it is valuable for tracking 3.x.
